### ISSUE TITLE
CUDA caching allocator can query an event recorded during thread-local capture

### ISSUE THREAD

**Issue by eee4017** [pytorch/pytorch#193982]
## Problem

During thread-local CUDA graph capture, PyTorch may free a default-pool block whose recorded use is on the capture stream. The native caching allocator records an event on that stream. A later allocation on the noncapturing default stream may poll the event and fail with `cudaErrorCapturedEvent` because the event was recorded during capture.

Minimal example, run in a fresh process:

```python
import torch

capture_stream = torch.cuda.Stream()
pending = torch.empty(1, device="cuda")
spare = torch.empty(1, device="cuda")
del spare  # keep one cached block available during capture

graph = torch.cuda.CUDAGraph()
with torch.cuda.stream(capture_stream):
    graph.capture_begin(capture_error_mode="thread_local")
    with torch.cuda.stream(torch.cuda.default_stream()):
        pending.record_stream(capture_stream)
        pending.untyped_storage().resize_(0)
        torch.empty(1, device="cuda")
    graph.capture_end()
```

No allocator configuration is required. The last allocation can enter ordinary event polling and query the event associated with `capture_stream` while capture is still active.

## Expected behavior

Allocator event maintenance should not record or query an event in a way that is invalid during capture. The block must remain protected until its recorded stream is safe.

## Possible Fix

A complete fix likely needs to track the stream for each event across `recordStream`, block release, event insertion, event polling, and forced cleanup. It must also coordinate event recording with capture start and end, handle overlapping captures, and continue processing events that belong only to safe streams.

Simply stopping all event processing while any capture is active would be smaller, but it could delay safe reclamation and cause avoidable memory pressure.

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng

**Comment by eqy** [comment]
What about simply adding a mechanism to force or wait until all pending eventqueries are gone? It sounds like the work is potentially incomplete if they are live? We used to do something similar for pending NCCL work that was done during e..g, the warmup https://github.com/pytorch/pytorch/pull/110665

**Comment by eee4017** [comment]
The problem is that, there are no pending allocator events at `capture_begin`. After capture starts, the default-stream free records a new event on the capture stream, and the following default-stream allocation queries that captured event.  We cannot wait for that event during capture because its graph node has not executed; CUDA reports `cudaErrorCapturedEvent`. 

A general allocator fix therefore needs to defer event insertion or querying based on the event’s target stream. It seems to me that both fixes are a little bit complex.

**Comment by eqy** [comment]
How does the free happen after the capture starts if the `del` happens before? Is it due to garbage collection?

**Comment by eqy** [comment]
Per offline discussion I think the current perspective is that deferring default-stream recycling due to this is fine as default-stream dependency/usage during graph capture is likely a user bug. We can avoid crashing in this case but don't need to treat it as an optimization priority.
